spring-pulsar/src/main/java/org/springframework/pulsar/listener/reactive/DefaultReactivePulsarMessageListenerContainer.java

```diff
@@ -0,0 +1,215 @@
+ * limitations under the License.
+ */
+
+package org.springframework.pulsar.listener.reactive;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.regex.Pattern;
+
+import org.apache.pulsar.client.api.Schema;
+import org.apache.pulsar.reactive.client.api.ReactiveMessageConsumer;
+import org.apache.pulsar.reactive.client.api.ReactiveMessagePipeline;
+import org.apache.pulsar.reactive.client.api.ReactiveMessagePipelineBuilder;
+import org.apache.pulsar.reactive.client.internal.api.ApiImplementationFactory;
+
+import org.springframework.core.log.LogAccessor;
+import org.springframework.pulsar.core.reactive.ReactiveMessageConsumerBuilderCustomizer;
+import org.springframework.pulsar.core.reactive.ReactivePulsarConsumerFactory;
+import org.springframework.util.Assert;
+
+/**
+ * Default implementation for {@link ReactivePulsarMessageListenerContainer}.
+ *
+ * @param <T> message type.
+ * @author Christophe Bornet
+ */
+public class DefaultReactivePulsarMessageListenerContainer<T> implements ReactivePulsarMessageListenerContainer {
+
+	private final LogAccessor logger = new LogAccessor(this.getClass());
+
+	private final ReactivePulsarConsumerFactory<T> pulsarConsumerFactory;
+
+	private final ReactivePulsarContainerProperties pulsarContainerProperties;
+
+	private boolean autoStartup = true;
+
+	private int phase;
+
+	private final Object lifecycleMonitor = new Object();
+
+	private volatile boolean running = false;
+
+	private ReactiveMessageConsumerBuilderCustomizer<T> consumerCustomizer;
+
+	private ReactiveMessagePipeline pipeline;
+
+	@SuppressWarnings("unchecked")
+	public DefaultReactivePulsarMessageListenerContainer(ReactivePulsarConsumerFactory<? super T> pulsarConsumerFactory,
+			ReactivePulsarContainerProperties pulsarContainerProperties) {
+		this.pulsarConsumerFactory = (ReactivePulsarConsumerFactory<T>) pulsarConsumerFactory;
+		this.pulsarContainerProperties = pulsarContainerProperties;
+	}
+
+	public ReactivePulsarConsumerFactory<T> getReactivePulsarConsumerFactory() {
+		return this.pulsarConsumerFactory;
+	}
+
+	public ReactivePulsarContainerProperties getContainerProperties() {
+		return this.pulsarContainerProperties;
+	}
+
+	@Override
+	public boolean isRunning() {
+		return this.running;
+	}
+
+	protected void setRunning(boolean running) {
+		this.running = running;
+	}
+
+	@Override
+	public void setupMessageHandler(Object messageHandler) {
+		this.pulsarContainerProperties.setMessageHandler(messageHandler);
+	}
+
+	@Override
+	public boolean isAutoStartup() {
+		return this.autoStartup;
+	}
+
+	@Override
+	public void setAutoStartup(boolean autoStartup) {
+		this.autoStartup = autoStartup;
+	}
+
+	public void setPhase(int phase) {
+		this.phase = phase;
+	}
+
+	@Override
+	public int getPhase() {
+		return this.phase;
+	}
+
+	@Override
+	public final void start() {
+		synchronized (this.lifecycleMonitor) {
```

Comment:
I would ❤️ to find a simpler guarded start/stop mechanism to use that leverages the `AtomicBoolean` rather than this "sync on monitor" approach. However, really we want to only allow one thing to start or stop at the same time. This would lead us to loop around `compareAndSet` which quickly becomes "not simpler" than this elder mechanism. Anyways, just thinking out loud.